### PR TITLE
Topic System Overhaul

### PR DESCRIPTION
**Pull request by MCterra10**
## About The Pull Request
Ike asked me really nicely so I reworked the whole topic system to use JSON and be less shit than before.
This PR ports a lot of Aurorastation/Aurora.3#825
and a little bit of Aurorastation/Aurora.3#876

### Request
Instead of passing weird arbitrary params, topic requests now require a simple JSON request:
```json
{
    "query": "api_function_key",
    "auth": "api_token",
    "source": "request source"
}
```
This is the absolute base request required for simple functions such as ping, status, playercount etc.
Many functions require additional arguments, though they're entirely reserved for internal functionality such as ahelps and crosscomms.

### Response
The server returns a simple JSON response in the following format:
```json
{
    "statuscode": 200,
    "response": "simple text information about the request's outcome",
    "data": "misc variable that can contain any additional information, such as a string or a list"
}
```
Any non-200 status code indicates an API error of some kind.

### Authentication
The authentication system has been completely overhauled in favor of a granular token permissions system.
An arbitrary amount of tokens can be created in the `comms.txt` config file, with specific permissions depending on usage.
See the config file for more elaborate configuration instructions.

***NOTE:*** By default, a static token exists which has the value of `anonymous`, and is used for unauthenticated topic calls. In the code, all topic datums have an anonymous flag, which if set to true, grants access to all tokens and the anonymous token by default.
This allows continued public access to topics such as status, playercount etc.

## Why It's Good For The Game
N/A; this changes nothing noticeable by players.

## Changelog
:cl: MCterra, Arrow768
server: Complete topic rework to use JSON and just be better in general
/:cl:

### CODE REVIEW

**Comment by MCterra10** [comment]
Note, I still need to account for not allowing some servers access to ahelp bridging. I'll sort that out soon.

Edit: going to add a handshake protocol to authenticate trusted servers

**Comment by ike709** [comment]
DNM until the bot and the website are also changed to be compatible with this.

**Comment by MCterra10** [comment]
alright so this is incredibly over-engineered but basically:
cross-server comms now requires a two-way "handshake" between both servers to tell each other what methods they have available and allow each server to work out what to send and what not to.

In simple terms, both servers need to have each other added as crosscomms servers, and only scopes that *they both have* are considered valid.
For example, for a server to forward ahelps to another server, it needs to verify that the other server is in possession of a key that can access ahelps on itself, and that it has a key to access ahelps on the other server.
 
Let me know if you want a terrible MS Paint flow chart of this lmao

**Comment by Crossedfall** [comment]
Work for this has been scheduled.

**Comment by MCterra10** [comment]
would be good if the code could get a pass over from a maintainer as well. I have tested it a bunch and stuff checks out, but a second pair of eyes would be good considering how core this stuff is.

**Comment by MCterra10** [comment]
Qwerty taking his time 😫

**Comment by utilbird** [comment]
See you next year

**Comment by MCterra10** [comment]
😔

**Review by Crossedfall — CHANGES_REQUESTED** [review]
Bot and website are pretty much good to go. This change needs actual documentation (i.e. a README) and conflict resolution before being ready for merge; otherwise, the major blockers are out of the way.

Can finally start going forward with this 🥳

**Comment by MCterra10** [comment]
Ugh I just noticed a bunch of the topic datum arg names aren't very consistent, I'll get this readme done and then fix that

**Comment by MCterra10** [comment]
could totally have synced asay between servers tbh

**Comment by MCterra10** [comment]
I also wouldn't mind writing some boilerplate code for targeted topic requests and to make things a bit more universal. Right now the comms_send proc only really works for ahelps, comms consoles and newscasters.

**Comment by MCterra10** [comment]
Code is now documented and I've made some nice improvements to the system in terms of code quality.
I'll run some tests on it tomorrow to make sure everything is in working order.

**Comment by MCterra10** [comment]
I've tested all topic endpoints (except ones that touch the DB) and made sure the responses are sane.
I've also tested cross-server comms; ahelps, comms consoles and newscasters are working.

File: code/datums/world_topic.dm
Comment (Crossedfall) [inline code/datums/world_topic.dm:73]: Is it a good idea to have a function that returns the full capabilities of a token? Seems a bit like a security nightmare.

Comment (ike709) [reply code/datums/world_topic.dm:73]: I mean I don't see what's stopping people from just attempting every API endpoint anyways.

Comment (Crossedfall) [reply code/datums/world_topic.dm:73]: Ratelimits stop them, mostly. It at least adds a significant enough delay to make it difficult to go through the entire list at once.

Comment (MCterra10) [reply code/datums/world_topic.dm:73]: I guess I can remove/disable it if you'd like, but I really don't see how getting authorised endpoints is a security issue. If you leaked a token and someone wants to use it maliciously, removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.

On the other hand, it's not really used for anything right now, since the only usecase within BYOND is already built into the handshake system.

All of that said, however, if you're developing an external tool, I think it's good to have a system to check what you can use, without having to actually poke the endpoints and see if you get 401 or not

Comment (Crossedfall) [reply code/datums/world_topic.dm:73]: >  removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.

That adds a pretty obvious pattern that can be observed. Even further, it'd be trivial to setup a monitor to watch for that kind of activity which can throw an alert for us to handle.

On one hand, I can see the value of this endpoint. On the other hand, I don't particularly like that it's an incredibly obvious attack vector.

Comment (Crossedfall) [reply code/datums/world_topic.dm:73]: I am not so committed to this that I'm willing to die on the hill though. So, if the general consensus is that it'll probably be fine, I'm willing to resign from the debate and we can move forward with it still openly available.

Comment (ike709) [reply code/datums/world_topic.dm:73]: It'll probably be fine

**Review by Crossedfall — APPROVED** [review]

**Comment by Crossedfall** [comment]
Staging tests seem good. Pending review from another maintainer before final merge

**Review by qwertyquerty — APPROVED** [review]

**Comment by KubeRoot** [comment]
This changes a lot of stuff, can somebody who knows this take a look at my PR and tell me how much was made redundant? https://github.com/BeeStation/BeeStation-Hornet/pull/4979

**Pull request merged**
